### PR TITLE
3.1 into 3.2

### PR DESCRIPTION
**Pull request by jack-w-shaw**
Forward merge:
- https://github.com/juju/juju/pull/16000
- https://github.com/juju/juju/pull/16004
- https://github.com/juju/juju/pull/16021
- https://github.com/juju/juju/pull/16020
- https://github.com/juju/juju/pull/16022  <-- All the conflicts from here
- https://github.com/juju/juju/pull/16025

There were many conflicts:
- agent/agentbootstrap/bootstrap.go
- api/client/imagemetadatamanager/client_test.go
- api/client/resources/client_upload_test.go
- apiserver/facades/client/bundle/bundle.go
- cmd/juju/application/deployer/bundlehandler.go
- cmd/juju/application/deployer/charm.go
- cmd/juju/application/deployer/charm_test.go
- cmd/juju/application/deployer/deployer.go
- cmd/juju/application/deployer/deployer_test.go
- cmd/juju/application/deployer/series_selector_test.go
- cmd/juju/commands/bootstrap_test.go
- cmd/plugins/juju-metadata/imagemetadata.go
- container/broker/broker.go
- container/broker/broker_test.go
- container/kvm/kvm.go
- container/kvm/live_test.go
- container/lxd/image.go
- container/lxd/manager.go
- core/charm/repository/charmhub.go
- core/charm/series_selector.go
- environs/bootstrap/bootstrap.go
- environs/bootstrap/bootstrap_test.go
- environs/bootstrap/tools.go
- environs/bootstrap/tools_test.go
- environs/imagedownloads/simplestreams.go
- environs/imagemetadata/simplestreams.go
- environs/jujutest/livetests.go
- environs/manual/sshprovisioner/sshprovisioner.go
- environs/sync/sync_test.go
- environs/tools/tools.go
- juju/testing/conn.go
- juju/testing/instance.go
- provider/azure/internal/imageutils/images.go
- provider/azure/internal/imageutils/images_test.go
- provider/common/bootstrap.go
- provider/common/bootstrap_test.go
- provider/dummy/environs.go
- provider/ec2/environ.go
- provider/ec2/local_test.go
- provider/equinix/environ.go
- provider/gce/environ_broker_test.go
- provider/gce/testing_test.go
- provider/lxd/environ_broker_test.go
- provider/lxd/testing_test.go
- provider/maas/environ.go
- provider/manual/environ.go
- provider/oci/images.go
- provider/oci/images_test.go
- provider/openstack/local_test.go
- provider/vsphere/environ_broker_test.go
- state/application_test.go
- state/state.go
- state/state_test.go
- testing/factory/factory.go

However, they were all fairly trivial

Most conflicts were purely in the imorts section. In which case, I simply removed the conflict marked andf let goimport remove unused packaged

Sometimes there were conflicts around the use of coreseries/corebase. In that case, I kept the code as it was in 3.2, and then changed coreseries to corebase

## QA Steps

```
make install
make static-analysis
```

### CODE REVIEW

File: state/state.go [hunk 1]
```diff
@@ -1457,14 +1457,14 @@ func (st *State) AddApplication(args AddApplicationArgs) (_ *Application, err er
 }
 
 func (st *State) processCommonModelApplicationArgs(args *AddApplicationArgs) (Base, error) {
-	// User has specified series. Overriding supported series is
+	// User has specified corebase. Overriding supported series is
```
Comment (hmlanigan) [inline state/state.go:1460]: I must not have caught this before.. s/corebase/base for this paragraph

File: state/state.go [hunk 2]
```diff
@@ -1168,7 +1168,7 @@ func (st *State) AddApplication(args AddApplicationArgs) (_ *Application, err er
 	}
 
 	// If either the charm origin ID or Hash is set before a charm is
-	// downloaded, charm download will fail for charms with a forced series.
+	// downloaded, charm download will fail for charms with a forced corebase.
```
Comment (hmlanigan) [inline state/state.go:1171]: s/corebase/base/

File: state/state.go [hunk 3]
```diff
@@ -1473,11 +1473,11 @@ func (st *State) processCommonModelApplicationArgs(args *AddApplicationArgs) (Ba
 	if cSeries := args.Charm.URL().Series; cSeries != "" {
 		supportedSeries = []string{cSeries}
 		// If a charm has a url, but is a kubernetes charm then we need to
-		// add this to the list of supported series.
-		if cSeries != series.Kubernetes.String() &&
-			(set.NewStrings(args.Charm.Meta().Series...).Contains(series.Kubernetes.String()) ||
+		// add this to the list of supported corebase.
```
Comment (hmlanigan) [inline state/state.go:1476]: s/corebase/base

File: apiserver/facades/client/application/application.go [hunk 4]
```diff
@@ -278,7 +278,7 @@ func (api *APIBase) Deploy(args params.ApplicationsDeploy) (params.ErrorResults,
 
 	for i, arg := range args.Applications {
 		// If either the charm origin ID or Hash is set before a charm is
-		// downloaded, charm download will fail for charms with a forced series.
+		// downloaded, charm download will fail for charms with a forced corebase.
```
Comment (hmlanigan) [inline apiserver/facades/client/application/application.go:281]: s/corebase/series

**Review by hmlanigan — COMMENTED** [review]
there may be more

**Review by cderici — APPROVED** [review]
This LGTM 👍

File: cmd/juju/application/refresh_test.go [hunk 5]
```diff
@@ -822,7 +822,7 @@ func (s *RefreshSuccessStateSuite) TestForcedSeriesUpgrade(c *gc.C) {
 	metadata := strings.Join(
 		[]string{
 			`name: multi-series`,
-			`summary: "That's a dummy charm with multi-series."`,
+			`summary: "That's a dummy charm with series."`,
```
Comment (cderici) [inline cmd/juju/application/refresh_test.go:825]: ☝️ https://github.com/juju/juju/pull/16030#discussion_r1280867979

Comment (jack-w-shaw) [reply cmd/juju/application/refresh_test.go:825]: Damn. Oh well, it's a arbitrary string in a test, not worth the effort I think

File: apiserver/facades/agent/upgradeseries/upgradeseries.go [hunk 6]
```diff
@@ -135,7 +135,7 @@ func (a *API) CurrentSeries(args params.Entities) (params.StringResults, error)
 			results[i].Error = apiservererrors.ServerError(err)
 			continue
 		}
-		series, err := coreseries.GetSeriesFromChannel(machine.Base().OS, machine.Base().Channel)
+		series, err := corebase.GetSeriesFromChannel(machine.Base().OS, machine.Base().Channel)
```
Comment (wallyworld) [inline apiserver/facades/agent/upgradeseries/upgradeseries.go:138]: We have plans to remove this and use base directly I assume. So this is just an interim step.

**Review by wallyworld — APPROVED** [review]
Awesome, ty

File: apiserver/facades/agent/upgradeseries/upgradeseries.go (hunk 6 above)
Comment (jack-w-shaw) [reply apiserver/facades/agent/upgradeseries/upgradeseries.go:138]: :+1:

**Pull request merged**